src/storage/prisma.ts

```diff
@@ -185,35 +165,77 @@ export class PrismaStorage implements IDriveStorage {
         operations: Operation[],
         header: DocumentHeader
     ): Promise<void> {
-        try {
-            await tx.operation.createMany({
-                data: operations.map(op => ({
-                    driveId: drive,
-                    documentId: id,
-                    hash: op.hash,
-                    index: op.index,
-                    input: JSON.stringify(op.input),
-                    timestamp: op.timestamp,
-                    type: op.type,
-                    scope: op.scope,
-                    branch: 'main',
-                    opId: op.id,
-                    skip: op.skip,
-                    context: op.context,
-                    resultingState: op.resultingState
-                        ? Buffer.from(JSON.stringify(op.resultingState))
-                        : undefined
-                }))
+        // TODO all operations should belong to the same sync unit
+        const groupedOperations = groupOperationsBySyncUnit(operations);
+        for (const { scope, branch, operations } of groupedOperations) {
+            const syncUnit = await tx.synchronizationUnit.findUnique({
+                where: {
+                    driveId_documentId_scope_branch: {
+                        driveId: drive,
+                        documentId: id,
+                        scope,
+                        branch
+                    }
+                },
+                select: {
+                    version: true
+                }
             });
+            if (!syncUnit) {
+                throw new Error(
+                    `Could not find sync unit for scope: ${JSON.stringify({ driveId: drive, documentId: id, scope, branch })}`
+                );
+            }
+            const revision = operations.reduce(
+                (acc, curr) => (curr.index > acc ? curr.index : acc),
+                -1
+            );
 
-            await tx.document.updateMany({
+            await tx.synchronizationUnit.update({
                 where: {
-                    id,
-                    driveId: drive
+                    driveId_documentId_scope_branch: {
+                        driveId: drive,
+                        documentId: id,
+                        scope,
+                        branch
+                    },
+                    version: syncUnit.version // optimistic update
```

Comment:
Adds the operations to the sync unit, including the version in the where clause, so that it will fail if another transaction incremented the version in between.